### DISCUSSION TITLE
ID navigation with offline creations

### DISCUSSION THREAD

**Discussion by Morzaram** [flutterdata/flutter_data#213] [Q&A]
Hey there I was wondering if there's a strategy to do ID navigation. Scenario:
App is offline first
1. user creates a new post 
2. Redirects the user to the post through a router `/posts/:id`
3. User hits button to sync data 
4. ???
5. Server and post have same ID 

If the post has a local only ID how do I save it safely to the server and have both IDs in sync?
My first thought is to make some sort of way where I replace the ID where on reconnect I delete ID, save and then replace it via `withKeyOf`. But I'm not sure if I'm over complicating things and theres an easier way. Trying to rack my brain around it. 

Reason why I'm doing this is that I want an offline first approach where people can only sync their data if they want. Ie: a click of a button. 

Thank you in advance as always ❤️

**Comment by Morzaram** [comment]
Edited for clarity

**Comment by frank06** [comment]
FD should automatically handle ID updates from a remote API:  https://github.com/flutterdata/flutter_data/blob/master/test/repository/repository_test.dart#L265

You could use UUIDs for offline/un-synced models. Once synced they accept whatever the server returned, could either (a) keep the UUID or (b) be replaced by an auto-increment . Up to you how to handle a potential redirect in the UI

**Reply by Morzaram** [reply]
Ah, okay, this makes a lot of sense! Thank you! 
So basically, on my server, if the ID doesn't exist, then I should make it if it's a PATCH request. Right? 
Or should I do it locally where I make a column with isOnServer = false. if isOnServer == false, then set to POST. 🤔

Thanks for your patience. I come from a SSR background and being this far away from the server is still so new to me

**Reply by frank06** [reply]
Easiest would be to work only with UUIDs. Server accepts whatever UUID the client has generated. If it does not exist write a DB row, if it exists update the row.

By default, if there is an ID present in the model the HTTP method will be `PATCH`, otherwise `POST`. You can override this via `methodForSave()` in your adapter (see docs). Same with `urlForSave()`

**Reply by frank06** [reply]
I also recommend you enable FD logging to get a better idea of what's going on. https://flutterdata.dev/docs/repositories/#logging-and-labels

**Reply by Morzaram** [reply]
Thank you so much for your help as always! If there's a way for me to donate or sponsor you? I want to compensate you some how.

**Reply by frank06** [reply]
No worries! Thank you
